File: sdk/python/inkbox/slack_operations.py
"""Scoped Slack utilities, durable actions, and explicitly retained history."""

from dataclasses import dataclass
from datetime import datetime
import re
from typing import Any, Literal
from urllib.parse import quote
from uuid import UUID

from inkbox._http import HttpTransport

SlackProcessingStatus = Literal["active", "processing", "suspended", "closed"]
SlackOperationKind = Literal[
    "reaction_add",
    "reaction_remove",
    "pin_add",
    "pin_remove",
    "message_update",
    "message_delete",
    "file_upload",
    "conversation_join",
    "conversation_leave",
    "processing_status",
]


@dataclass
class SlackOperation:
    id: UUID
    connection_id: UUID
    operation: SlackOperationKind
    status: Literal["in_progress", "succeeded", "failed", "unknown"]
    conversation_id: str
    message_ts: str | None = None
    file_id: str | None = None
    error_code: str | None = None
    retry_after: int | None = None
    processing_status: SlackProcessingStatus | None = None
    agent_status: SlackProcessingStatus | None = None


@dataclass
class SlackCapability:
    required_scopes: list[str]
    missing_scopes: list[str]
    scopes_satisfied: bool


@dataclass
class SlackCapabilitiesResponse:
    connection_id: UUID
    scopes: list[str]
    missing_scopes: list[str]
    capabilities: dict[str, SlackCapability]
    native_processing_status: Literal["unknown", "missing_scope"]
    max_upload_bytes: int


@dataclass
class SlackUsersResponse:
    users: list[dict[str, Any]]
    next_cursor: str | None = None


@dataclass
class SlackMembersResponse:
    members: list[str]
    next_cursor: str | None = None


@dataclass
class SlackPinsResponse:
    items: list[dict[str, Any]]
    truncated: bool = False


@dataclass
class SlackReactionsResponse:
    conversation_id: str
    message_ts: str
    reactions: list[dict[str, Any]]


@dataclass
class SlackMessageContextResponse:
    messages: list[dict[str, Any]]
    next_cursor: str | None = None
    has_more: bool = False
    window: Literal["messages_at_or_before_timestamp"] = (
        "messages_at_or_before_timestamp"
    )
    complete: Literal[False] = False


@dataclass
class SlackPermalinkResponse:
    conversation_id: str
    message_ts: str
    permalink: str


@dataclass
class SlackArchiveSettings:
    capture_enabled: bool
    retention_days: int | None
    conversation_ids: list[str]
    revision: int


@dataclass
class SlackArchivedMessage:
    id: UUID
    connection_id: UUID
    conversation_id: str
    message_ts: str
    thread_ts: str | None
    user_id: str | None
    text: str
    files: list[dict[str, Any]]
    mentioned: bool
    source: Literal["event", "backfill", "action"]
    captured_at: datetime
    source_url: str | None = None


@dataclass
class SlackArchiveMessagesResponse:
    messages: list[SlackArchivedMessage]
    next_cursor: str | None = None
    source: Literal["archive"] = "archive"


@dataclass
class SlackArchiveCoverage:
    conversation_id: str
    thread_ts: str | None
    status: Literal["observed", "pending", "running", "complete", "failed", "paused"]
    imported_count: int
    oldest_ts: str | None
    newest_ts: str | None
    error_code: str | None
    updated_at: datetime
    includes_all_threads: bool = False
    access_revoked: bool = False


@dataclass
class SlackArchiveCoverageResponse:
    coverage: list[SlackArchiveCoverage]
    next_cursor: str | None = None


def _parse(cls, raw):
    values = {
        key: value for key, value in raw.items() if key in cls.__dataclass_fields__
    }
    for key in ("id", "connection_id"):
        if key in values:
            values[key] = UUID(values[key])
    for key in ("captured_at", "updated_at"):
        if key in values:
            values[key] = datetime.fromisoformat(values[key])
    return cls(**values)


def _base(value: UUID | str) -> str:
    return f"/slack/connections/{quote(str(value), safe='')}"


def _conversation(connection_id: UUID | str, conversation_id: str) -> str:
    return f"{_base(connection_id)}/conversations/{quote(conversation_id, safe='')}"


def _timestamp(value: str | None) -> str | None:
    if value is not None and not isinstance(value, str):
        raise ValueError("Slack timestamps must be strings")
    return value


def _message(connection_id: UUID | str, conversation_id: str, message_ts: str) -> str:
    _timestamp(message_ts)
    return f"{_conversation(connection_id, conversation_id)}/messages/{quote(message_ts, safe='')}"


class SlackOperationsMixin:
    _http: HttpTransport

    def _operation(
        self, method: str, path: str, key: str, body: dict | None = None
    ) -> SlackOperation:
        if not isinstance(key, str) or not re.fullmatch(r"[A-Za-z0-9._:-]{1,128}", key):
            raise ValueError(
                "idempotency_key must be 1..128 letters, digits, '.', '_', ':', or '-'"
            )
        kwargs = {"headers": {"Idempotency-Key": key}}
        if method != "delete_with_response":
            kwargs["json"] = body
        return _parse(SlackOperation, getattr(self._http, method)(path, **kwargs))

    def capabilities(self, connection_id: UUID | str) -> SlackCapabilitiesResponse:
        raw = self._http.get(f"{_base(connection_id)}/capabilities")
        raw["capabilities"] = {
            name: _parse(SlackCapability, value)
            for name, value in raw["capabilities"].items()
        }
        return _parse(SlackCapabilitiesResponse, raw)

    def list_users(
        self, connection_id: UUID | str, *, limit: int = 100, cursor: str | None = None
    ) -> SlackUsersResponse:
        return _parse(
            SlackUsersResponse,
            self._http.get(
                f"{_base(connection_id)}/users",
                params={"limit": limit, "cursor": cursor},
            ),
        )

    def get_user(self, connection_id: UUID | str, user_id: str) -> dict[str, Any]:
        return self._http.get(f"{_base(connection_id)}/users/{quote(user_id, safe='')}")

    def list_members(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        *,
        limit: int = 100,
        cursor: str | None = None,
    ) -> SlackMembersResponse:
        return _parse(
            SlackMembersResponse,
            self._http.get(
                f"{_conversation(connection_id, conversation_id)}/members",
                params={"limit": limit, "cursor": cursor},
            ),
        )

    def get_message(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        *,
        thread_ts: str | None = None,
    ) -> dict[str, Any]:
        return self._http.get(
            _message(connection_id, conversation_id, message_ts),
            params={"thread_ts": _timestamp(thread_ts)},
        )

    def message_context(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        *,
        limit: int = 5,
        thread_ts: str | None = None,
    ) -> SlackMessageContextResponse:
        return _parse(
            SlackMessageContextResponse,
            self._http.get(
                f"{_message(connection_id, conversation_id, message_ts)}/context",
                params={"limit": limit, "thread_ts": _timestamp(thread_ts)},
            ),
        )

    def get_permalink(
        self, connection_id: UUID | str, conversation_id: str, message_ts: str
    ) -> SlackPermalinkResponse:
        return _parse(
            SlackPermalinkResponse,
            self._http.get(
                f"{_message(connection_id, conversation_id, message_ts)}/permalink"
            ),
        )

    def get_reactions(
        self, connection_id: UUID | str, conversation_id: str, message_ts: str
    ) -> SlackReactionsResponse:
        return _parse(
            SlackReactionsResponse,
            self._http.get(
                f"{_message(connection_id, conversation_id, message_ts)}/reactions"
            ),
        )

    def list_pins(
        self, connection_id: UUID | str, conversation_id: str
    ) -> SlackPinsResponse:
        return _parse(
            SlackPinsResponse,
            self._http.get(f"{_conversation(connection_id, conversation_id)}/pins"),
        )

    def get_operation(
        self, connection_id: UUID | str, operation_id: UUID | str
    ) -> SlackOperation:
        """Poll only in_progress; unknown is terminal uncertainty, not a retry instruction."""
        return _parse(
            SlackOperation,
            self._http.get(
                f"{_base(connection_id)}/operations/{quote(str(operation_id), safe='')}"
            ),
        )

    def add_reaction(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        name: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        return self._operation(
            "post",
            f"{_message(connection_id, conversation_id, message_ts)}/reactions",
            idempotency_key,
            {"name": name},
        )

    def remove_reaction(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        name: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        return self._operation(
            "delete_with_response",
            f"{_message(connection_id, conversation_id, message_ts)}/reactions/{quote(name, safe='')}",
            idempotency_key,
        )

    def add_pin(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        return self._operation(
            "post",
            f"{_conversation(connection_id, conversation_id)}/pins",
            idempotency_key,
            {"message_ts": _timestamp(message_ts)},
        )

    def remove_pin(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        _timestamp(message_ts)
        return self._operation(
            "delete_with_response",
            f"{_conversation(connection_id, conversation_id)}/pins/{quote(message_ts, safe='')}",
            idempotency_key,
        )

    def update_message(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        text: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        """Edit only the connected agent's own message; Slack enforces authorship."""
        return self._operation(
            "patch",
            _message(connection_id, conversation_id, message_ts),
            idempotency_key,
            {"text": text},
        )

    def delete_message(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        message_ts: str,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        return self._operation(
            "delete_with_response",
            _message(connection_id, conversation_id, message_ts),
            idempotency_key,
        )

    def join_conversation(
        self, connection_id: UUID | str, conversation_id: str, *, idempotency_key: str
    ) -> SlackOperation:
        return self._operation(
            "post",
            f"{_conversation(connection_id, conversation_id)}/join",
            idempotency_key,
        )

    def leave_conversation(
        self, connection_id: UUID | str, conversation_id: str, *, idempotency_key: str
    ) -> SlackOperation:
        return self._operation(
            "post",
            f"{_conversation(connection_id, conversation_id)}/leave",
            idempotency_key,
        )

    def set_processing_status(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        thread_ts: str,
        status: SlackProcessingStatus,
        *,
        idempotency_key: str,
    ) -> SlackOperation:
        """Native feature availability is workspace-dependent; no reaction fallback."""
        return self._operation(
            "post",
            f"{_conversation(connection_id, conversation_id)}/processing-status",
            idempotency_key,
            {"thread_ts": _timestamp(thread_ts), "status": status},
        )

    def upload_file(
        self,
        connection_id: UUID | str,
        *,
        conversation_id: str,
        filename: str,
        content_base64: str,
        idempotency_key: str,
        title: str | None = None,
        initial_comment: str | None = None,
        thread_ts: str | None = None,
    ) -> SlackOperation:
        """Upload 1 byte..10 MiB of general file content encoded as standard base64."""
        body = {
            "conversation_id": conversation_id,
            "filename": filename,
            "content_base64": content_base64,
        }
        body.update(
            {
                key: value
                for key, value in {
                    "title": title,
                    "initial_comment": initial_comment,
                    "thread_ts": _timestamp(thread_ts),
                }.items()
                if value is not None
            }
        )
        return self._operation(
            "post", f"{_base(connection_id)}/files", idempotency_key, body
        )

    def get_archive_settings(self, connection_id: UUID | str) -> SlackArchiveSettings:
        return _parse(
            SlackArchiveSettings,
            self._http.get(f"{_base(connection_id)}/archive/settings"),
        )

    def update_archive_settings(
        self,
        connection_id: UUID | str,
        *,
        capture_enabled: bool,
        retention_days: int | None = None,
        conversation_ids: list[str] | None = None,
    ) -> SlackArchiveSettings:
        """Organization management only; replaces capture settings. Null retention has no time limit."""
        return _parse(
            SlackArchiveSettings,
            self._http.patch(
                f"{_base(connection_id)}/archive/settings",
                json={
                    "capture_enabled": capture_enabled,
                    "retention_days": retention_days,
                    "conversation_ids": conversation_ids
                    if conversation_ids is not None
                    else [],
                },
            ),
        )

    def _archive_messages(
        self, connection_id: UUID | str, path: str, params: dict
    ) -> SlackArchiveMessagesResponse:
        for key in ("thread_ts", "before_ts", "after_ts"):
            _timestamp(params.get(key))
        raw = self._http.get(f"{_base(connection_id)}/archive/{path}", params=params)
        raw["messages"] = [
            _parse(SlackArchivedMessage, item) for item in raw["messages"]
        ]
        return _parse(SlackArchiveMessagesResponse, raw)

    def list_archived_messages(
        self,
        connection_id: UUID | str,
        *,
        conversation_id: str | None = None,
        thread_ts: str | None = None,
        before_ts: str | None = None,
        after_ts: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> SlackArchiveMessagesResponse:
        return self._archive_messages(
            connection_id,
            "messages",
            {
                "conversation_id": conversation_id,
                "thread_ts": thread_ts,
                "before_ts": before_ts,
                "after_ts": after_ts,
                "cursor": cursor,
                "limit": limit,
            },
        )

    def search_archived_messages(
        self,
        connection_id: UUID | str,
        q: str,
        *,
        conversation_id: str | None = None,
        user_id: str | None = None,
        before_ts: str | None = None,
        after_ts: str | None = None,
        cursor: str | None = None,
        limit: int = 50,
    ) -> SlackArchiveMessagesResponse:
        return self._archive_messages(
            connection_id,
            "search",
            {
                "q": q,
                "conversation_id": conversation_id,
                "user_id": user_id,
                "before_ts": before_ts,
                "after_ts": after_ts,
                "cursor": cursor,
                "limit": limit,
            },
        )

    def archive_backfill(
        self,
        connection_id: UUID | str,
        conversation_id: str,
        *,
        thread_ts: str | None = None,
        restart: bool = False,
    ) -> SlackArchiveCoverage:
        body = {"conversation_id": conversation_id, "restart": restart}
        if thread_ts is not None:
            body["thread_ts"] = _timestamp(thread_ts)
        return _parse(
            SlackArchiveCoverage,
            self._http.post(f"{_base(connection_id)}/archive/backfill", json=body),
        )

    def list_archive_coverage(
        self, connection_id: UUID | str, *, limit: int = 100, cursor: str | None = None
    ) -> SlackArchiveCoverageResponse:
        raw = self._http.get(
            f"{_base(connection_id)}/archive/coverage",
            params={"limit": limit, "cursor": cursor},
        )
        raw["coverage"] = [
            _parse(SlackArchiveCoverage, item) for item in raw["coverage"]
        ]
        return _parse(SlackArchiveCoverageResponse, raw)

    def purge_archive(self, connection_id: UUID | str) -> dict[str, Any]:
        """Organization management only; disables capture and queues retained-content deletion."""
        return self._http.delete_with_response(f"{_base(connection_id)}/archive")

File: sdk/typescript/src/slack-operations.ts
/** Scoped Slack utilities, durable actions, and retained history. */
import type { HttpTransport } from "./_http.js";
import type { SlackPageOptions } from "./slack.js";

export type SlackProcessingStatus =
  "active" | "processing" | "suspended" | "closed";
export type SlackOperationKind =
  | "reaction_add"
  | "reaction_remove"
  | "pin_add"
  | "pin_remove"
  | "message_update"
  | "message_delete"
  | "file_upload"
  | "conversation_join"
  | "conversation_leave"
  | "processing_status";
export interface SlackOperation {
  id: string;
  connectionId: string;
  operation: SlackOperationKind;
  status: "in_progress" | "succeeded" | "failed" | "unknown";
  conversationId: string;
  messageTs: string | null;
  fileId: string | null;
  errorCode: string | null;
  retryAfter: number | null;
  processingStatus: SlackProcessingStatus | null;
  agentStatus: SlackProcessingStatus | null;
}
export interface SlackCapability {
  requiredScopes: string[];
  missingScopes: string[];
  scopesSatisfied: boolean;
}
export interface SlackCapabilitiesResponse {
  connectionId: string;
  scopes: string[];
  missingScopes: string[];
  capabilities: Record<string, SlackCapability>;
  nativeProcessingStatus: "unknown" | "missing_scope";
  maxUploadBytes: number;
}
export interface SlackUsersResponse {
  users: Record<string, unknown>[];
  nextCursor: string | null;
}
export interface SlackMembersResponse {
  members: string[];
  nextCursor: string | null;
}
export interface SlackPinsResponse {
  items: Record<string, unknown>[];
  truncated: boolean;
}
export interface SlackReactionsResponse {
  conversationId: string;
  messageTs: string;
  reactions: Record<string, unknown>[];
}
export interface SlackMessageContextResponse {
  messages: Record<string, unknown>[];
  nextCursor: string | null;
  hasMore: boolean;
  window: "messages_at_or_before_timestamp";
  complete: false;
}
export interface SlackPermalinkResponse {
  conversationId: string;
  messageTs: string;
  permalink: string;
}
export interface SlackArchiveSettings {
  captureEnabled: boolean;
  retentionDays: number | null;
  conversationIds: string[];
  revision: number;
}
export interface SlackArchiveSettingsOptions {
  captureEnabled: boolean;
  retentionDays?: number | null;
  conversationIds?: string[];
}
export interface SlackArchivedMessage {
  id: string;
  connectionId: string;
  conversationId: string;
  messageTs: string;
  threadTs: string | null;
  userId: string | null;
  text: string;
  files: Record<string, unknown>[];
  mentioned: boolean;
  source: "event" | "backfill" | "action";
  sourceUrl: string | null;
  capturedAt: Date;
}
export interface SlackArchiveMessagesResponse {
  messages: SlackArchivedMessage[];
  nextCursor: string | null;
  source: "archive";
}
export interface SlackArchiveCoverage {
  conversationId: string;
  threadTs: string | null;
  status: "observed" | "pending" | "running" | "complete" | "failed" | "paused";
  importedCount: number;
  oldestTs: string | null;
  newestTs: string | null;
  errorCode: string | null;
  updatedAt: Date;
  includesAllThreads: boolean;
  accessRevoked: boolean;
}
export interface SlackArchiveCoverageResponse {
  coverage: SlackArchiveCoverage[];
  nextCursor: string | null;
}
export interface SlackArchiveMessagesOptions extends SlackPageOptions {
  conversationId?: string;
  threadTs?: string | null;
  beforeTs?: string | null;
  afterTs?: string | null;
}
export interface SlackArchiveSearchOptions extends Omit<
  SlackArchiveMessagesOptions,
  "threadTs"
> {
  userId?: string;
}
export interface SlackUploadFileOptions {
  conversationId: string;
  filename: string;
  contentBase64: string;
  idempotencyKey: string;
  title?: string | null;
  initialComment?: string | null;
  threadTs?: string | null;
}
export interface SlackMutationOptions {
  idempotencyKey: string;
}

type Snake<S extends string> = S extends `${infer A}${infer B}`
  ? `${A extends Lowercase<A> ? A : `_${Lowercase<A>}`}${Snake<B>}`
  : S;
type Wire<T> = {
  [K in keyof T as Snake<K & string>]: T[K] extends Date ? string : T[K];
};
type RawOperation = Wire<SlackOperation>;
const base = (id: string): string =>
  `/slack/connections/${encodeURIComponent(id)}`;
const conversation = (id: string, channel: string): string =>
  `${base(id)}/conversations/${encodeURIComponent(channel)}`;
function timestamp(
  value: string | null | undefined,
): string | null | undefined {
  if (value != null && typeof value !== "string")
    throw new TypeError("Slack timestamps must be strings");
  return value;
}
const message = (id: string, channel: string, ts: string): string => {
  timestamp(ts);
  return `${conversation(id, channel)}/messages/${encodeURIComponent(ts)}`;
};
const operation = (r: RawOperation): SlackOperation => ({
  id: r.id,
  connectionId: r.connection_id,
  operation: r.operation,
  status: r.status,
  conversationId: r.conversation_id,
  messageTs: r.message_ts ?? null,
  fileId: r.file_id ?? null,
  errorCode: r.error_code ?? null,
  retryAfter: r.retry_after ?? null,
  processingStatus: r.processing_status ?? null,
  agentStatus: r.agent_status ?? null,
});
const settings = (r: Wire<SlackArchiveSettings>): SlackArchiveSettings => ({
  captureEnabled: r.capture_enabled,
  retentionDays: r.retention_days,
  conversationIds: r.conversation_ids,
  revision: r.revision,
});
const archivedMessage = (
  r: Wire<SlackArchivedMessage>,
): SlackArchivedMessage => ({
  id: r.id,
  connectionId: r.connection_id,
  conversationId: r.conversation_id,
  messageTs: r.message_ts,
  threadTs: r.thread_ts,
  userId: r.user_id,
  text: r.text,
  files: r.files,
  mentioned: r.mentioned,
  source: r.source,
  sourceUrl: r.source_url ?? null,
  capturedAt: new Date(r.captured_at),
});
const coverage = (r: Wire<SlackArchiveCoverage>): SlackArchiveCoverage => ({
  conversationId: r.conversation_id,
  threadTs: r.thread_ts,
  status: r.status,
  importedCount: r.imported_count,
  oldestTs: r.oldest_ts,
  newestTs: r.newest_ts,
  errorCode: r.error_code,
  updatedAt: new Date(r.updated_at),
  includesAllThreads: r.includes_all_threads ?? false,
  accessRevoked: r.access_revoked ?? false,
});

export class SlackOperationsResource {
  constructor(protected readonly http: HttpTransport) {}
  private async mutate(
    method: "POST" | "PATCH" | "DELETE",
    path: string,
    key: string,
    body?: unknown,
  ): Promise<SlackOperation> {
    if (typeof key !== "string" || !/^[A-Za-z0-9._:-]{1,128}$/.test(key))
      throw new TypeError(
        "idempotencyKey must be 1..128 letters, digits, '.', '_', ':', or '-'",
      );
    const opts = { headers: { "Idempotency-Key": key } };
    const result =
      method === "DELETE"
        ? await this.http.deleteWithResponse<RawOperation>(path, opts)
        : method === "PATCH"
          ? await this.http.patch<RawOperation>(path, body, opts)
          : await this.http.post<RawOperation>(path, body, opts);
    return operation(result);
  }
  async capabilities(connectionId: string): Promise<SlackCapabilitiesResponse> {
    const r = await this.http.get<
      Omit<Wire<SlackCapabilitiesResponse>, "capabilities"> & {
        capabilities: Record<string, Wire<SlackCapability>>;
      }
    >(`${base(connectionId)}/capabilities`);
    return {
      connectionId: r.connection_id,
      scopes: r.scopes,
      missingScopes: r.missing_scopes,
      nativeProcessingStatus: r.native_processing_status,
      maxUploadBytes: r.max_upload_bytes,
      capabilities: Object.fromEntries(
        Object.entries(r.capabilities).map(([name, value]) => [
          name,
          {
            requiredScopes: value.required_scopes,
            missingScopes: value.missing_scopes,
            scopesSatisfied: value.scopes_satisfied,
          },
        ]),
      ),
    };
  }
  async listUsers(
    connectionId: string,
    options: SlackPageOptions = {},
  ): Promise<SlackUsersResponse> {
    const r = await this.http.get<Wire<SlackUsersResponse>>(
      `${base(connectionId)}/users`,
      { limit: options.limit ?? 100, cursor: options.cursor },
    );
    return { users: r.users, nextCursor: r.next_cursor ?? null };
  }
  async getUser(
    connectionId: string,
    userId: string,
  ): Promise<Record<string, unknown>> {
    return this.http.get(
      `${base(connectionId)}/users/${encodeURIComponent(userId)}`,
    );
  }
  async listMembers(
    connectionId: string,
    conversationId: string,
    options: SlackPageOptions = {},
  ): Promise<SlackMembersResponse> {
    const r = await this.http.get<Wire<SlackMembersResponse>>(
      `${conversation(connectionId, conversationId)}/members`,
      { limit: options.limit ?? 100, cursor: options.cursor },
    );
    return { members: r.members, nextCursor: r.next_cursor ?? null };
  }
  async getMessage(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    options: { threadTs?: string | null } = {},
  ): Promise<Record<string, unknown>> {
    return this.http.get(message(connectionId, conversationId, messageTs), {
      thread_ts: timestamp(options.threadTs),
    });
  }
  async messageContext(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    options: { threadTs?: string | null; limit?: number } = {},
  ): Promise<SlackMessageContextResponse> {
    const r = await this.http.get<Wire<SlackMessageContextResponse>>(
      `${message(connectionId, conversationId, messageTs)}/context`,
      { thread_ts: timestamp(options.threadTs), limit: options.limit ?? 5 },
    );
    return {
      messages: r.messages,
      nextCursor: r.next_cursor ?? null,
      hasMore: r.has_more ?? false,
      window: r.window,
      complete: r.complete,
    };
  }
  async getPermalink(
    connectionId: string,
    conversationId: string,
    messageTs: string,
  ): Promise<SlackPermalinkResponse> {
    const r = await this.http.get<Wire<SlackPermalinkResponse>>(
      `${message(connectionId, conversationId, messageTs)}/permalink`,
    );
    return {
      conversationId: r.conversation_id,
      messageTs: r.message_ts,
      permalink: r.permalink,
    };
  }
  async getReactions(
    connectionId: string,
    conversationId: string,
    messageTs: string,
  ): Promise<SlackReactionsResponse> {
    const r = await this.http.get<Wire<SlackReactionsResponse>>(
      `${message(connectionId, conversationId, messageTs)}/reactions`,
    );
    return {
      conversationId: r.conversation_id,
      messageTs: r.message_ts,
      reactions: r.reactions,
    };
  }
  async listPins(
    connectionId: string,
    conversationId: string,
  ): Promise<SlackPinsResponse> {
    return this.http.get(`${conversation(connectionId, conversationId)}/pins`);
  }
  /** Poll only in_progress. Unknown is terminal uncertainty; never blindly repeat. */
  async getOperation(
    connectionId: string,
    operationId: string,
  ): Promise<SlackOperation> {
    return operation(
      await this.http.get(
        `${base(connectionId)}/operations/${encodeURIComponent(operationId)}`,
      ),
    );
  }
  async addReaction(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    name: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${message(connectionId, conversationId, messageTs)}/reactions`,
      options.idempotencyKey,
      { name },
    );
  }
  async removeReaction(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    name: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "DELETE",
      `${message(connectionId, conversationId, messageTs)}/reactions/${encodeURIComponent(name)}`,
      options.idempotencyKey,
    );
  }
  async addPin(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${conversation(connectionId, conversationId)}/pins`,
      options.idempotencyKey,
      { message_ts: timestamp(messageTs) },
    );
  }
  async removePin(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    timestamp(messageTs);
    return this.mutate(
      "DELETE",
      `${conversation(connectionId, conversationId)}/pins/${encodeURIComponent(messageTs)}`,
      options.idempotencyKey,
    );
  }
  /** Slack enforces that the connected agent can only edit its own message. */
  async updateMessage(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    text: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "PATCH",
      message(connectionId, conversationId, messageTs),
      options.idempotencyKey,
      { text },
    );
  }
  async deleteMessage(
    connectionId: string,
    conversationId: string,
    messageTs: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "DELETE",
      message(connectionId, conversationId, messageTs),
      options.idempotencyKey,
    );
  }
  async joinConversation(
    connectionId: string,
    conversationId: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${conversation(connectionId, conversationId)}/join`,
      options.idempotencyKey,
    );
  }
  async leaveConversation(
    connectionId: string,
    conversationId: string,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${conversation(connectionId, conversationId)}/leave`,
      options.idempotencyKey,
    );
  }
  /** Native workspace support is not implied by scopes; unsupported calls remain explicit failures. */
  async setProcessingStatus(
    connectionId: string,
    conversationId: string,
    threadTs: string,
    status: SlackProcessingStatus,
    options: SlackMutationOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${conversation(connectionId, conversationId)}/processing-status`,
      options.idempotencyKey,
      { thread_ts: timestamp(threadTs), status },
    );
  }
  /** Standard base64 of any file type, bounded to 1 byte..10 MiB decoded. */
  async uploadFile(
    connectionId: string,
    options: SlackUploadFileOptions,
  ): Promise<SlackOperation> {
    return this.mutate(
      "POST",
      `${base(connectionId)}/files`,
      options.idempotencyKey,
      {
        conversation_id: options.conversationId,
        filename: options.filename,
        content_base64: options.contentBase64,
        title: options.title,
        initial_comment: options.initialComment,
        thread_ts: timestamp(options.threadTs),
      },
    );
  }
  async getArchiveSettings(
    connectionId: string,
  ): Promise<SlackArchiveSettings> {
    return settings(
      await this.http.get(`${base(connectionId)}/archive/settings`),
    );
  }
  /** Organization management only. Replaces capture settings; null retention has no time limit. */
  async updateArchiveSettings(
    connectionId: string,
    options: SlackArchiveSettingsOptions,
  ): Promise<SlackArchiveSettings> {
    return settings(
      await this.http.patch(`${base(connectionId)}/archive/settings`, {
        capture_enabled: options.captureEnabled,
        retention_days: options.retentionDays ?? null,
        conversation_ids: options.conversationIds ?? [],
      }),
    );
  }
  private async archiveMessages(
    connectionId: string,
    path: string,
    options: SlackArchiveMessagesOptions,
    extra: Record<string, string | undefined> = {},
  ): Promise<SlackArchiveMessagesResponse> {
    const r = await this.http.get<{
      messages: Wire<SlackArchivedMessage>[];
      next_cursor?: string | null;
      source: "archive";
    }>(`${base(connectionId)}/archive/${path}`, {
      conversation_id: options.conversationId,
      thread_ts: timestamp(options.threadTs),
      before_ts: timestamp(options.beforeTs),
      after_ts: timestamp(options.afterTs),
      cursor: options.cursor,
      limit: options.limit ?? 50,
      ...extra,
    });
    return {
      messages: r.messages.map(archivedMessage),
      nextCursor: r.next_cursor ?? null,
      source: r.source,
    };
  }
  async listArchivedMessages(
    connectionId: string,
    options: SlackArchiveMessagesOptions = {},
  ): Promise<SlackArchiveMessagesResponse> {
    return this.archiveMessages(connectionId, "messages", options);
  }
  async searchArchivedMessages(
    connectionId: string,
    q: string,
    options: SlackArchiveSearchOptions = {},
  ): Promise<SlackArchiveMessagesResponse> {
    return this.archiveMessages(connectionId, "search", options, {
      q,
      user_id: options.userId,
    });
  }
  async archiveBackfill(
    connectionId: string,
    conversationId: string,
    options: { threadTs?: string | null; restart?: boolean } = {},
  ): Promise<SlackArchiveCoverage> {
    return coverage(
      await this.http.post(`${base(connectionId)}/archive/backfill`, {
        conversation_id: conversationId,
        thread_ts: timestamp(options.threadTs),
        restart: options.restart ?? false,
      }),
    );
  }
  async listArchiveCoverage(
    connectionId: string,
    options: SlackPageOptions = {},
  ): Promise<SlackArchiveCoverageResponse> {
    const r = await this.http.get<{
      coverage: Wire<SlackArchiveCoverage>[];
      next_cursor?: string | null;
    }>(`${base(connectionId)}/archive/coverage`, {
      limit: options.limit ?? 100,
      cursor: options.cursor,
    });
    return {
      coverage: r.coverage.map(coverage),
      nextCursor: r.next_cursor ?? null,
    };
  }
  /** Organization management only; disables capture and queues retained-content deletion. */
  async purgeArchive(
    connectionId: string,
  ): Promise<{ status: "pending"; captureEnabled: false }> {
    const r = await this.http.deleteWithResponse<{
      status: "pending";
      capture_enabled: false;
    }>(`${base(connectionId)}/archive`);
    return { status: r.status, captureEnabled: r.capture_enabled };
  }
}
